discord/ext/commands/bot.py

```diff
@@ -568,20 +568,31 @@ def _call_module_finalizers(self, lib, key):
                 if _is_submodule(name, module):
                     del sys.modules[module]
 
-    def _load_from_module_spec(self, lib, key):
+    def _load_from_module_spec(self, spec, key):
         # precondition: key not in self.__extensions
+        def unload(e):
+            self._remove_module_references(lib.__name__)
+            self._call_module_finalizers(lib, key)
+            raise errors.ExtensionFailed(key, e) from e
+
+        lib = importlib.util.module_from_spec(spec)
+        try:
+            spec.loader.exec_module(lib)
+        except Exception as e:
+            unload(e)
+
         try:
             setup = getattr(lib, 'setup')
         except AttributeError:
-            del sys.modules[key]
             raise errors.NoEntryPointError(key)
 
+        sys.modules[key] = lib
```

Comment:
Why is this being assigned here if you're going to delete it later anyway? Just add it when everything is all done.